### PR TITLE
🌱 Add unit test cases for clustermodule

### PR DESCRIPTION
**Pull request by MrDXY**
**What this PR does / why we need it**:
1. Add unit test cases for `pkg/clustermodule/service.go`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2308

### CODE REVIEW

File: pkg/session/fake/fake_session.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (zhanggbj) [inline pkg/session/fake/fake_session.go:2]: It should be 2023

File: pkg/session/fake/fake_session.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+func MockDefaultSession() (*session.Session, error) {
+	return MockSession(DefaultDataCenter)
+}
+
+// MockSession GetOrCreate gets a cached session or creates a new one if one does not.
+func MockSession(dc string) (*session.Session, error) {
+	model := simulator.VPX()
+	model.Cluster = 1
```
Comment (zhanggbj) [inline pkg/session/fake/fake_session.go:46]: The model.Cluster returned by VPX() is already `1`.

File: pkg/session/fake/fake_session.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	Server   = ""
+	Username = ""
+	Password = ""
+)
+
+func MockDefaultSession() (*session.Session, error) {
+	return MockSession(DefaultDataCenter)
+}
+
+// MockSession GetOrCreate gets a cached session or creates a new one if one does not.
```
Comment (zhanggbj) [inline pkg/session/fake/fake_session.go:43]:
```suggestion
// MockSession gets a cached session or creates a new one if one does not already exist.
```

File: Makefile [hunk 4]
```diff
@@ -454,7 +454,7 @@ setup-envtest: $(SETUP_ENVTEST) ## Set up envtest (download kubebuilder assets)
 
 .PHONY: test
 test: $(SETUP_ENVTEST) $(GOVC) ## Run unit tests
-	KUBEBUILDER_ASSETS="$(KUBEBUILDER_ASSETS)" GOVC_BIN_PATH=$(GOVC) go test -v ./apis/... ./controllers/... ./pkg/... ./internal/... $(TEST_ARGS)
```
Comment (chrischdi) [inline Makefile:457]: Why is that needed?

File: pkg/clustermodule/service_test.go [hunk 5]
```diff
@@ -21,13 +21,25 @@ import (
 	"fmt"
 	"testing"
 
+	"github.com/agiledragon/gomonkey/v2"
```
Comment (chrischdi) [inline pkg/clustermodule/service_test.go:24]: Question: did you consider to use `github.com/stretchr/testify/mock` instead because we already use that in other tests.
I just want to make sure to introduce a second variant of mocking stuff.

File: pkg/session/fake/fake_session.go (hunk 1 above)
Comment (MrDXY) [reply pkg/session/fake/fake_session.go:2]: Yes, of course.

File: pkg/session/fake/fake_session.go (hunk 2 above)
Comment (MrDXY) [reply pkg/session/fake/fake_session.go:46]: Noted, I will remove this statement.

File: pkg/clustermodule/service_test.go (hunk 5 above)
Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: Sure, I have considered using `/testify/mock`. 
1. The general idea of using mock in go has two pre-requirements/constraints:
    1. Have interfaces that allow you to switch to your mocked behavior.
    2. Inject a mocked struct in place of the real one(dependency injection).
2. But that is not always the case, especially when:
    1. Working on a code base that is unalterable.
    2. Lacks programming to interface, only structs, hindering dependency injection.
3. In our case:
    1. I found it hard to mock a `session`'s behavior. Especially when attempting to obtain a session with `govmomi` client. Certain internal calls, such as the call to `s.SessionManager.UserSession(ctx)`, could not be mocked by `testify`.
    2. Also the same for other `govmomi` client calls.
4. So I decided to use `gomonkey`, a tool that utilizes go to modify functions at runtime.
    1. With this mocking toolkit, we can easily mock the behavior of target functions. Like modifying the return objects and redefining the function execution process.
    2. And here I use it to mock some of the functions of `govmomi` client.

File: Makefile (hunk 4 above)
Comment (MrDXY) [reply Makefile:457]: That is because `gomonkey` doesn't support inline functions, and during testing, we need to disable Go language's inline optimization by using the command-line argument `-gcflags=-l`.

File: pkg/clustermodule/service_test.go [hunk 6]
```diff
@@ -81,6 +93,172 @@ func TestService_Create(t *testing.T) {
 			g.Expect(moduleUUID).To(gomega.BeEmpty())
 		})
 	})
+
+	t.Run("creation is executed", func(t *testing.T) {
+		g := gomega.NewWithT(t)
+		_, err := fake_session.MockDefaultSession()
```
[the remark this answers is not in the document]
Comment (MrDXY) [reply pkg/clustermodule/service_test.go:99]: We'll use it [here](https://github.com/MrDXY/cluster-api-provider-vsphere/blob/pr_unit_test_for_clustermodule/pkg/clustermodule/service_test.go#L146C1-L147) (when you try to create a cluster module)
https://github.com/MrDXY/cluster-api-provider-vsphere/blob/4d2c987b67375a803280b248ef361469d5f870f6/pkg/clustermodule/service_test.go#L146C1-L147

My consideration here is to cache a session and use it for upcoming tests rather than create a session when using it. But it's also an option, only execute `vcsim.NewBuilder()...`, and let the case create a session.

File: pkg/clustermodule/service_test.go (hunk 5 above)
Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: Also, if you are concerned about the license issue, `agiledragon/monkey` is licensed under the MIT License. Our work on increasing ut-coverage could benefit from this, especially when mocking the interaction with `govmomi`.

Comment (chrischdi) [reply pkg/clustermodule/service_test.go:24]: I'm not concerned about the license. I just lean towards trying to prevent additional dependencies when there are alternatives. The only project in the k8s ecosystem which uses gomonkey is [vsphere-csi](https://cs.k8s.io/?q=github.com%2Fagiledragon%2Fgomonkey&i=nope&files=&excludeFiles=&repos=). 

All functionality should also be possible to get tested using vcsim as backend, I currently see no need to use monkey patching.

Another point is disabling code optimisations during tests. If we disable code optimisations we do test compiled code which is not the very same as what we ship.

Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: > All functionality should also be possible to get tested using vcsim as backend

I attempted to utilize vcsim to establish a testing environment previously, but I encountered an error. 
- Error when I try to create a cluster module:
    - The error message indicates that vcenter does not have the requested path, namely `POST https://127.0.0.1:64753/rest/vcenter/cluster/modules`.image
- Also the same 404 error for `ListModules`.
- Have you encounter similar errors? May be I got this wrong, but that's literally the reason that I found it hard to mock a govmomi client's behavior. Maybe we can not simulate all vCenter's responses via vcsim.

As for code optimizations during tests, that's just an inner mechanism of Go, it should work identically with or without those optimizations. The result should be the same, just an efficiency issue. No need to be worried : ）

Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: > I just lean towards trying to prevent additional dependencies when there are alternatives

It also makes sense to me, but one dependency should be OK, and it is only used for test purposes.

Comment (chrischdi) [reply pkg/clustermodule/service_test.go:24]: What I see is:

I can run vcsim and use it to create/list/remove cluster modules.

```sh
❯ vcsim &
export GOVC_URL=https://user:pass@127.0.0.1:8989/sdk GOVC_SIM_PID=39024
❯ export GOVC_URL=https://user:pass@127.0.0.1:8989/sdk GOVC_SIM_PID=39024
❯ export GOVC_INSECURE=true
❯ govc cluster.module.ls
❯ govc cluster.module.create
d5a10942-7178-4f9c-996c-cead7e8b4490
❯ govc cluster.module.ls
domain-c27  d5a10942-7178-4f9c-996c-cead7e8b4490
❯ govc cluster.module.rm d5a10942-7178-4f9c-996c-cead7e8b4490
❯ govc cluster.module.ls
```

> The result should be the same, just an efficiency issue. 

Yeah, but still a small risk that it behaves differently.

Comment (killianmuldoon) [reply pkg/clustermodule/service_test.go:24]: 100 percent agreed with @chrischdi on this - we should use vcsim for this testing if at all possible.

Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: Got it, folks. Let me take a closer look at vcsim and try to figure out & solve the problem I met : )

Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: Update: still encounter the same 404 error while using the command line with `vcsim.Simulator` (I'll go deeper into this issue.) Also, please feel free to share any of your ideas or suggestions on this matter.
image

Comment (MrDXY) [reply pkg/clustermodule/service_test.go:24]: I think I have identified the cause of the 404 error I encountered. The cluster module request will be dealt with [here](https://github.com/vmware/govmomi/blob/244e3ac4bfc31edacd09f191e6a5411ad0c15064/vapi/cluster/simulator/simulator.go#L71-L108). The [init](https://github.com/vmware/govmomi/blob/244e3ac4bfc31edacd09f191e6a5411ad0c15064/vapi/cluster/simulator/simulator.go#L37-L41) function is used to register the handler/endpoint to the simulator, however, the function will not execute if we only import `github.com/vmware/govmomi/simulator`. To ensure the `init` function works, it is necessary to import `github.com/vmware/govmomi/vapi/cluster/simulator` as well. 

Here is the code for the `init` function:
```
func init() {
        simulator.RegisterEndpoint(func(s *simulator.Service, r *simulator.Registry) {
                New(s.Listen).Register(s, r)
        })
}
```

After importing the required package, everything appears to be working properly. I will update the corresponding changes accordingly!

File: pkg/clustermodule/service_test.go [hunk 7]
```diff
@@ -81,6 +84,148 @@ func TestService_Create(t *testing.T) {
...
+		controllerCtx := fake.NewControllerContext(fake.NewControllerManagerContext(md, machineTemplate))
+		clusterCtx := fake.NewClusterContext(controllerCtx)
+		clusterCtx.VSphereCluster.Spec.Server = simr.ServerURL().Host
+		controllerCtx.ControllerManagerContext.Username = simr.Username()
+		controllerCtx.ControllerManagerContext.Password = simr.Password()
+
+		svc := NewService(controllerCtx.ControllerManagerContext, controllerCtx.Client)
+		moduleUUID, err := svc.Create(context.Background(), clusterCtx, mdWrapper{md})
+		g.Expect(err).ToNot(gomega.HaveOccurred())
+		g.Expect(moduleUUID).NotTo(gomega.BeEmpty())
```
Comment (chrischdi) [inline pkg/clustermodule/service_test.go:126]: I think it's not worth having a separated `DoesExist` and `Delete` test functions.

These two also test teh `Create` part already. Let's combine them to a single test.


```suggestion
		g.Expect(moduleUUID).NotTo(gomega.BeEmpty())


		exists, err := svc.DoesExist(context.Background(), clusterCtx, mdWrapper{md}, moduleUUID)
		g.Expect(exists).To(gomega.BeTrue())
		g.Expect(err).NotTo(gomega.HaveOccurred())

		err = svc.Remove(context.Background(), clusterCtx, moduleUUID)
		g.Expect(err).ToNot(gomega.HaveOccurred())
```

**Review by chrischdi — COMMENTED** [review]
NIce work with using vcims. This gets very clean :-)

File: pkg/clustermodule/service_test.go [hunk 8]
```diff
@@ -81,6 +84,148 @@ func TestService_Create(t *testing.T) {
 			g.Expect(moduleUUID).To(gomega.BeEmpty())
 		})
 	})
+
+	t.Run("creation is executed", func(t *testing.T) {
+		g := gomega.NewWithT(t)
+		simr, err := initVCSimulator()
```
[the remark this answers is not in the document]
Comment (MrDXY) [reply pkg/clustermodule/service_test.go:90]: That makes sense. Let me remove this and make the necessary adjustments.

**Comment by MrDXY** [comment]
Hi @chrischdi , This PR is ready for another round of review; please feel free to take a look whenever you get a chance. 😄

**Comment by chrischdi** [comment]
> Hi @chrischdi , This PR is ready for another round of review; please feel free to take a look whenever you get a chance. 😄

Sorry for the late reply, forgot to send out the above comment 🤦

**Comment by MrDXY** [comment]
> Sorry for the late reply, forgot to send out the above comment 🤦

No problem at all. I'll take care of this. 😄

File: pkg/clustermodule/service_test.go [hunk 9]
```diff
@@ -81,6 +82,52 @@ func TestService_Create(t *testing.T) {
 			g.Expect(moduleUUID).To(gomega.BeEmpty())
 		})
 	})
+
+	t.Run("creation is executed", func(t *testing.T) {
```
Comment (chrischdi) [inline pkg/clustermodule/service_test.go:86]:
```suggestion
	t.Run("Create, DoesExist and Remove works", func(t *testing.T) {
```

**Review by chrischdi — COMMENTED** [review]
Last nit

/lgtm

**Review by chrischdi — COMMENTED** [review]
/approve

**Review by killianmuldoon — COMMENTED** [review]
/lgtm
/approve

Thanks!

**Review by chrischdi — COMMENTED** [review]
Only nits about comments and a optional question/thing for the test code, lgtm to it in general!

Thanks for also adding test cases!
